pkg/storage/ingest/pusher.go

```diff
@@ -241,163 +252,281 @@ func (c pusherConsumer) shouldLogClientError(ctx context.Context, err error) (bo
-		s.Timeseries = append(s.Timeseries, ts)
-		s.Context = ctx // retain the last context in case we have to flush it when closing shardingPusher
-
-		if len(s.Timeseries) < p.batchSize {
-			continue
-		}
-		p.unfilledShards[shard] = flushableWriteRequest{
-			WriteRequest: &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()},
+		// TODO: Add metrics to measure how long are items sitting in the queue before they are flushed.
+		if err := p.shards[shard].AddToBatch(ctx, ts); err != nil {
+			errs.Add(err)
 		}
+	}
 
-	tryPush:
-		for {
-			select {
-			case p.shards[shard] <- s:
-				break tryPush
-			case err := <-p.errs:
-				// we might have to first unblock the shard before we can flush to it
-				errs.Add(err)
-			}
-		}
+	// We might some data left in some of the queues in the shards, but they will be flushed eventually once Stop is called, and we're certain that no more data is coming.
+	// Return whatever errors we have now, we'll call stop eventually and collect the rest.
+	return errs.Err()
+}
 
-		// drain errors to avoid blocking the shard loop
-	drainErrs:
-		for {
-			select {
-			case err := <-p.errs:
-				errs.Add(err)
-			default:
-				break drainErrs
-			}
-		}
+// Stop stops all the shards and waits for them to finish.
+func (p *parallelStorageShards) Stop() error {
+	var errs multierror.MultiError
+
+	for _, shard := range p.shards {
+		errs.Add(shard.Close())
 	}
+
+	p.wg.Wait()
+
 	return errs.Err()
 }
 
-type flushableWriteRequest struct {
-	*mimirpb.WriteRequest
-	context.Context
+// start starts the shards, each in its own goroutine.
+func (p *parallelStorageShards) start() {
+	shards := make([]*BatchingQueue, p.numShards)
+	p.wg.Add(p.numShards)
+
+	for i := range shards {
+		shards[i] = NewBatchingQueue(p.capacity, p.batchSize)
+		go p.run(shards[i])
+	}
+
+	p.shards = shards
 }
 
-func (p *shardingPusher) runShard(toFlush chan flushableWriteRequest) {
+// run runs the BatchingQueue for the shard.
+func (p *parallelStorageShards) run(queue *BatchingQueue) {
 	defer p.wg.Done()
-	for wr := range toFlush {
+	defer queue.Done()
+
+	for wr := range queue.Channel() {
 		p.numTimeSeriesPerFlush.Observe(float64(len(wr.WriteRequest.Timeseries)))
-		err := p.upstream.PushToStorage(wr.Context, wr.WriteRequest)
+		err := p.pusher.PushToStorage(wr.Context, wr.WriteRequest)
 		if err != nil {
-			p.errs <- err
+			queue.ErrorChannel() <- err
+		}
+	}
+}
+
+// BatchingQueue is a queue that batches the incoming time series according to the batch size.
+// Once the batch size is reached, the batch is pushed to a channel which can be accessed through the Channel() method.
+type BatchingQueue struct {
+	ch    chan FlushableWriteRequest
+	errCh chan error
+	done  chan struct{}
+
+	currentBatch FlushableWriteRequest
+	batchSize    int
+}
+
+// NewBatchingQueue creates a new BatchingQueue instance.
+func NewBatchingQueue(capacity int, batchSize int) *BatchingQueue {
```

Comment:
do these need to be exported? I wouldn't be opposed ot moving them to a different package though. Something like `pkg/ingest/push`? If not, we can unexport them. Same for `FlushableWriteRequest`

Reply:
Done in https://github.com/grafana/mimir/pull/9133/commits/7e56c2f2af28a31ac6f4e1dc832f4d89e287b583

I made them private, I don't think they warrant a different package / file just yet.